### PR TITLE
converted arrays to objects

### PR DESCRIPTION
**Pull request by darrachequesne**
See #2199

### CODE REVIEW

**Comment by rauchg** [comment]
Wouldn't it be better to use a `Map` here?

**Comment by nkzawa** [comment]
Object would be better for now, since there is no `Map` on Node.js 0.10.

**Comment by darrachequesne** [comment]
@nkzawa yes, that's my opinion too :+1: 

If we'd want to use  ̀Map` (and more generally the ES2015 features) and still keep support of node < 0.12, I think a migration to Babel (or other) would be necessary (but that's out of the scope here)

**Comment by rauchg** [comment]
In the meantime we can merge this.
Moving forward, I think we should be using Babel and `babel-runtime` + runtime transformer to provide `Map`.

**Comment by rauchg** [comment]
Please rebase. Current status: image

**Comment by darrachequesne** [comment]
Done! :angel:

File: lib/index.js [hunk 1]
```diff
@@ -347,9 +347,9 @@ Server.prototype.of = function(name, fn){
  */
 
 Server.prototype.close = function(){
-  this.nsps['/'].sockets.forEach(function(socket){
-    socket.onclose();
-  });
+  for (var id in this.nsps['/'].sockets) {
```
Comment (rauchg) [inline lib/index.js]: You're not checking for `hasOwnProperty` here. Extensions fo the natives will make socket.io break here. Test needed for that scenario as well.

File: lib/namespace.js [hunk 2]
```diff
@@ -187,9 +187,8 @@ Namespace.prototype.add = function(client, fn){
  */
 
 Namespace.prototype.remove = function(socket){
-  var i = this.sockets.indexOf(socket);
-  if (~i) {
-    this.sockets.splice(i, 1);
+  if (this.sockets[socket.id]) {
```
Comment (rauchg) [inline lib/namespace.js]: `this.sockets.hasOwnProperty` would be a better test. 
See for inspiration: https://github.com/socketio/socket.io-adapter/pull/30

**Comment by darrachequesne** [comment]
Hi! I opened a new PR to clean this ugly one: https://github.com/socketio/socket.io/pull/2322

**Pull request merged**